Sidebar transparency effect

mj-shifu opened the pull request:
<!--

    Add a concise description of what this PR is changing or adding, and why. Consider including before/after screenshots.
    Consider mentioninig issues related to this pull request

-->

This PR adds a transparency effect to the sidebar, as long as system and app brightness match.

Some changes to MainFlutterWindow.swift are necessary, like they were done in the README and the example

## Pre-launch Checklist

- [x] I have run `dartfmt` on all changed files <!-- THIS IS REQUIRED -->
- [x] I have incremented the package version as appropriate and updated `CHANGELOG.md` with my changes <!-- THIS IS REQUIRED -->
- [x] I have added/updated relevant documentation <!-- If relevant -->
- [x] I have run "optimize/organize imports" on all changed files
- [x] I have addressed all analyzer warnings as best I could
<!-- - [ ] I have run `flutter pub publish --dry-run` and addressed any warnings --> <!-- MAINTAINER ONLY -->

### Before PR
<img width="842" alt="before_white" src="https://user-images.githubusercontent.com/77107165/149658031-9f69171e-7757-49ff-a52a-edcd00e65289.png">
<img width="842" alt="before_black" src="https://user-images.githubusercontent.com/77107165/149658036-10837b45-c028-4b1c-820a-939d2a8ae63a.png">

## After PR
<img width="842" alt="after_light" src="https://user-images.githubusercontent.com/77107165/149658040-099b145c-d4d5-4cab-a34e-ae95e03b974d.png">
<img width="842" alt="after_black" src="https://user-images.githubusercontent.com/77107165/149658044-7055aa06-2e2f-4f3d-a613-8c79480a1a8e.png">

mj-shifu committed:
Transparency sidebar effect

Added a transparency effect to the sidebar

mj-shifu committed:
fix: Formatted window.dart

mj-shifu committed:
chore: update changelog, version

mj-shifu committed:
fix: removed unused imports

mj-shifu commented:
This should close #173.

jmatth commented on lib/src/layout/window.dart:
```diff
@@ -95,7 +95,12 @@ class _MacosWindowState extends State<MacosWindow> {
     late Color sidebarBackgroundColor;
     Color dividerColor = theme.dividerColor;
 
-    if (!theme.brightness.isDark) {
+    // Only show blurry, transparent sidebar when platform brightness and app
+    // brightness are the same, otherwise it looks awful
+    if (MediaQuery.of(context).platformBrightness.isDark ==
+        theme.brightness.isDark) {
+      sidebarBackgroundColor = Colors.transparent;
+    } else if (!theme.brightness.isDark) {
       sidebarBackgroundColor = widget.sidebar?.decoration?.color ??
           CupertinoColors.systemGrey6.color;
     } else {
```
2 things:

- This doesn't look right on web and Windows (I didn't test Linux). Screenshots below.
- I would expect any color override from the parent to take precedence, but right now it will be overridden by the transparency effect.

I would suggest the following changes to fix both issues:

```suggestion
    final isMac = !kIsWeb && defaultTargetPlatform == TargetPlatform.macOS;
    
    // Respect the sidebar color override from parent if one is given
    if (widget.sidebar?.decoration?.color != null) {
      sidebarBackgroundColor = widget.sidebar!.decoration!.color!;
    } else if (isMac &&
        MediaQuery.of(context).platformBrightness.isDark ==
            theme.brightness.isDark) {
      // Only show blurry, transparent sidebar when platform brightness and app
      // brightness are the same, otherwise it looks awful. Also only make the
      // sidebar transparent on native Mac, or it will just be flat black or
      // white.
      sidebarBackgroundColor = Colors.transparent;
    } else if (!theme.brightness.isDark) {
      sidebarBackgroundColor = widget.sidebar?.decoration?.color ??
          CupertinoColors.systemGrey6.color;
    } else {
      sidebarBackgroundColor = widget.sidebar?.decoration?.color ??
          CupertinoColors.tertiarySystemBackground.darkColor;
}
```

This change is meant to replace line 97 - 109 rather than 97 - 106, but Github won't let me apply the suggestion that far outside the diff 😒.

Web:

![image](https://user-images.githubusercontent.com/1316184/150372855-2a2dc59b-e507-46b3-9380-83d660bcd318.png)

Windows:

![image](https://user-images.githubusercontent.com/1316184/150372660-dd9cbe71-d424-47bc-94b0-4b60590fbf8b.png)

jmatth reviewed (changes requested):
Caveat on my review: I am not a member of this repo and can't merge. Also thanks for doing this, you just saved me digging through some Apple docs I was not looking forward to 😁.

mj-shifu committed:
fix: applied enhancement from @jmatth

this fixes weird behavior on windows and web

mj-shifu commented:
Thank you, I just changed it according to your suggestion.

jmatth commented on lib/src/layout/window.dart:
```diff
@@ -95,7 +96,20 @@ class _MacosWindowState extends State<MacosWindow> {
+      // sidebar transparent on native Mac, or it will just be flat black or
+      // white.
+      sidebarBackgroundColor = Colors.transparent;
+    } else if (!theme.brightness.isDark) {
```
Sorry to but in again, I just realized the suggestion I sent was an older version of the code. Since we're now checking for `widget.sidebar?.decoration?.color` in the first `if` branch, we don't need to consider it when falling back to the theme. The last two branches can be collapsed into single else clause with a ternary statement:

```suggestion
    } else {
        sidebarBackgroundColor = CupertinoColors.systemGrey6.color;
    }
```

Or if you prefer without the ternary:

```suggestion
    } else if (!theme.brightness.isDark) {
        sidebarBackgroundColor = CupertinoColors.systemGrey6.color;
    } else {
        sidebarBackgroundColor = CupertinoColors.tertiarySystemBackground.darkColor;
    }
```

Both changes are supposed to replace lines 112 - 118 and haven't been run through `dartfmt` because github's suggestion system is very annoying to use.

GroovinChip replied:
@mj-shifu what are your thoughts on @jmatth 's feedback here?

mj-shifu replied:
I agree with @jmatth 's feedback. But because GitHub wouldn't let me merge the changes directly, I created a new commit to fix it.

mj-shifu committed:
fix: double null check

widget.sidebar.decoration.color was checked twice
(according to suggestion by @jmatth)

GroovinChip reviewed (approved):
LGTM! Thanks for your contribution, @mj-shifu !
